pkg/hubble/cell/hubbleintegration.go

```diff
@@ -341,16 +341,49 @@ func (h *hubbleIntegration) launch(ctx context.Context) {
 
 	var srv *http.Server
 	if h.config.MetricsServer != "" {
-		h.log.WithFields(logrus.Fields{
-			"address": h.config.MetricsServer,
-			"metrics": h.config.Metrics,
-			"tls":     h.config.EnableMetricsServerTLS,
-		}).Info("Starting Hubble Metrics server")
+		switch {
+		case h.config.DynamicMetricConfigFilePath != "" && len(h.config.Metrics) > 0:
+			{
+				h.log.Error("Cannot configure both static and dynamic Hubble metrics")
+			}
+		case h.config.DynamicMetricConfigFilePath != "":
+			{
+				h.log.WithFields(logrus.Fields{
+					"address":      h.config.MetricsServer,
+					"metricConfig": h.config.DynamicMetricConfigFilePath,
+					"tls":          h.config.EnableMetricsServerTLS,
+				}).Info("Starting Hubble server with dynamically configurable metrics")
+
+				metrics.InitHubbleInternalMetrics(metrics.Registry, grpcMetrics)
+				dynamicFp := metrics.NewDynamicFlowProcessor(metrics.Registry, h.log, h.config.DynamicMetricConfigFilePath)
+				observerOpts = append(observerOpts, observeroption.WithOnDecodedFlow(dynamicFp))
+			}
+		case len(h.config.Metrics) > 0:
+			{
+				h.log.WithFields(logrus.Fields{
+					"address": h.config.MetricsServer,
+					"metrics": h.config.Metrics,
+					"tls":     h.config.EnableMetricsServerTLS,
+				}).Info("Starting Hubble Metrics server")
+
+				err := metrics.InitMetrics(metrics.Registry, api.ParseStaticMetricsConfig(h.config.Metrics), grpcMetrics)
+				if err != nil {
+					h.log.WithError(err).Error("Unable to setup metrics: %w", err)
+					return
+				}
 
-		err := metrics.InitMetrics(metrics.Registry, api.ParseStaticMetricsConfig(h.config.Metrics), grpcMetrics)
-		if err != nil {
-			h.log.WithError(err).Error("Unable to setup metrics: %w", err)
-			return
+				observerOpts = append(observerOpts,
+					observeroption.WithOnDecodedFlowFunc(func(ctx context.Context, flow *flowpb.Flow) (bool, error) {
+						if metrics.EnabledMetrics != nil {
```

Comment:
This is the static path, but this if-clause can’t be moved to the switch condition because metrics.EnabledMetrics is only initialized by metrics.InitMetrics() within this case block.
You could argue that the callback registered by OnDecodedFlowFunc will never encounter a nil value, but I left the check there unchanged.

Reply:
Note, this section of the code is unchanged by this PR, other than moving out the contents of the pass-through `ProcessFlow` method on the now-removed wrapper type `Handlers`.

Reply:
Note that this section of the code is unchanged by this PR, other than removing the wrapper function over the now-redundant wrapper type's `Handlers.ProcessFlow`.
```
func ProcessFlow(ctx context.Context, flow *pb.Flow) error {
	if enabledMetrics != nil {
		return enabledMetrics.ProcessFlow(ctx, flow)
	}
	return nil
}
```